test/smoke_test.go

```diff
@@ -0,0 +1,314 @@
+	}
+
+	return nil
+}
+
+func waitForDeployment(ctx context.Context, c client.Client) error {
+	var continuousUptime int
+	return eventually(func() error {
+		dl := &appsv1.DeploymentList{}
+		if err := c.List(ctx, dl, &client.ListOptions{Namespace: scannerNamespace}); err != nil {
+			return fmt.Errorf("error listing deployments: %w", err)
+		}
+		if len(dl.Items) != 1 {
+			continuousUptime = 0
+			return fmt.Errorf("expected 1 deployment, but got %v", len(dl.Items))
+		}
+		deploy := dl.Items[0]
+		if deploy.Status.AvailableReplicas < 1 {
+			continuousUptime = 0
+			return fmt.Errorf("no replicas available")
+		}
+		continuousUptime++
+		if continuousUptime > 10 {
+			return nil
+
+		}
+		return fmt.Errorf("not enough continuous uptime yet: %v", continuousUptime)
+	}, 2*time.Minute, time.Second)
+}
+
+func eventually(fn func() error, timeout time.Duration, tick time.Duration) error {
+	after := time.After(timeout)
+	ticker := time.NewTicker(tick)
+
+	var err error
+	if err = fn(); err == nil {
+		return nil
+	}
+	for {
+		select {
+		case <-after:
+			return fmt.Errorf("timed out waiting for condition. last error: %w", err)
+		case <-ticker.C:
+			err = fn()
+			if err == nil {
+				return nil
+			}
+		}
+	}
+}
+
+type helmValues struct {
+	Config config.Config `json:"config"`
+}
+
+func newHelmValues(t *testing.T) (vals helmValues, filename string) {
+	orgID := os.Getenv("TEST_ORGANIZATION_ID")
+	if orgID == "" {
+		t.Fatalf("no TEST_ORGANIZATION_ID env var set")
+	}
+
+	snykSAToken := os.Getenv("TEST_SNYK_SERVICE_ACCOUNT_TOKEN")
+	if snykSAToken == "" {
+		t.Fatalf("TEST_SNYK_SERVICE_ACCOUNT_TOKEN not specified!")
+	}
+
+	vals = helmValues{
+		Config: config.Config{
+			ClusterName: newClusterName(),
+			// TODO: test more complex routing?
+			Routes: []config.Route{{
+				ClusterScopedResources: true,
+				Namespaces:             []string{"*"},
+				OrganizationID:         orgID,
+			}},
+			Scanning: config.Scan{
+				Types: []config.ScanType{{
+					APIGroups: []string{""},
+					Versions:  []string{"v1"},
+					Resources: []string{"nodes"},
+				}},
+				RequeueAfter: metav1.Duration{Duration: time.Hour},
+			},
+			Egress: &config.Egress{
+				SnykAPIBaseURL:          "https://api.dev.snyk.io",
+				SnykServiceAccountToken: snykSAToken,
+			},
+		},
+	}
+
+	values, err := json.Marshal(vals)
+	if err != nil {
+		t.Fatalf("could not render routing config: %v", err)
+	}
+
+	f, err := os.CreateTemp(t.TempDir(), "routes_values_*.yaml")
+	if err != nil {
+		t.Errorf("could not write temporary values.yaml file: %v", err)
+	}
+
+	if _, err := f.Write(values); err != nil {
+		t.Errorf("could not write routes to temporary file: %v", err)
+	}
+
+	f.Close()
+
+	return vals, f.Name()
+}
+
+func installTilt(t *testing.T) {
+	t.Helper()
+	// use the current GOROOT to pick which Go to build with
+	goBin := filepath.Join(build.Default.GOROOT, "bin", "go")
+	cmd := exec.Command(goBin, "install", "github.com/tilt-dev/tilt/cmd/tilt")
```

Comment:
I wasn't sure what to do with this to be honest.
Right now we have  a split: we install `kubectl`, `helm`, `kind` and `ctlptl` in CI, and `tilt` when the test starts to run.

All versions are defined through the `go.mod` file as we use `go install` to get them, except for `kubectl` where we fetch a binary (I at least ensured that it's the same version as our k8s libs).

Setup of the cluster happens in CI, setup of Tilt happens in the test.

The split primarily happened because we want this to work on dev's machines as well. The current setup is "require a kubernetes cluster" (connected to through the kubeconfig file), and `tilt` also requires `kubectl` and `helm` to be present.

Should we drop the `go install tilt` in this test and move it to CI? This _could_ mean that devs have a different version of tilt installed, but I'm not sure if that matters much.

With that, we'd require all tooling to be present already when `go test`ing; a Kubernetes cluster, `kubectl`, `helm` and `tilt`. I _think_ this is reasonable, but I'm open to other opinions & suggestions!

Reply:
I removed this for now and install `tilt` in CI as well. It's clearer than doing it in the code, which is a weird side effect.